core/src/main/java/org/apache/druid/math/expr/ExprMacroTable.java

```diff
@@ -97,9 +99,13 @@ public Expr get(final String functionName, final List<Expr> args)
   {
     protected final Expr arg;
 
+    // Use Supplier to memoize values as ExpressionSelectors#makeExprEvalSelector() can make repeated calls for them
+    private final Supplier<BindingDetails> analyzeInputsSupplier;
+
     public BaseScalarUnivariateMacroFunctionExpr(Expr arg)
     {
       this.arg = arg;
+      analyzeInputsSupplier = Suppliers.memoize(this::supplyAnalyzeInputs);
```

Comment:
do we want this to happen at construction time always or this needs to be lazily memoized ?

Reply:
What do you mean by lazily memoize? Do you mean to lazily instantiate the `Supplier`?

Reply:
sorry, ignore , its already lazy.

Reply:
btw, did the perf difference come from this change or from converting streams to for loops ?